### PR TITLE
dtoh: Add explicit conversions for not implictly convertible expressions

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
Adds special cases for several implicit conversions that do not translate directly to C++ (slices initialized to null, strings, ...)

### CODE REVIEW

File: src/dmd/dtoh.d [hunk 1]
```diff
@@ -652,10 +652,11 @@ public:
 
                 case EnumKind.String, EnumKind.Enum:
                     buf.writestring("static ");
-                    writeEnumTypeName(type);
+                    auto target = determineEnumType(type);
+                    target.accept(this);
                     buf.printf(" const %s = ", vd.ident.toChars());
```
Comment (thewilsonator) [inline src/dmd/dtoh.d:657]: `determineEnumType(memberType).accept(this);`

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/dtoh.d (hunk 1 above)
Comment (MoonlightSentinel) [reply src/dmd/dtoh.d:657]: `target` is used below and `memberType` is not defined here?

Comment (thewilsonator) [reply src/dmd/dtoh.d:657]: my bad I missed that.

**Comment by MoonlightSentinel** [comment]
CC @ibuclaw

File: test/compilable/dtoh_cpp98_compat.d [hunk 2]
```diff
@@ -48,14 +50,32 @@ private:
 public:
     Null() :
         field(NULL),
-        null_(_d_dynamicArray< const char >())
+        null_()
```
Comment (kinke) [inline test/compilable/dtoh_cpp98_compat.d:53]: Thx, much better.

**Comment by UplinkCoder** [comment]
This still looks a bit too complicated for my liking.
But I wouldn't oppose it.

**Comment by UplinkCoder** [comment]
`determineEnumType` is always followed by `this.accept()`
as far as I can see.
Why not keep it as writeEnumType ?

**Comment by MoonlightSentinel** [comment]
> Why not keep it as writeEnumType ?

Because then I would need to add more special cases to detect & handle type substitutions (which would duplicate the logic in `weriteEnumType`).
Using a real type instance to makes `printExpressionFor` (and probably some other code in the future) work as is.

File: src/dmd/dtoh.d [hunk 3]
```diff
@@ -1855,7 +1856,107 @@ public:
...
+            // writefln("%s:%s: target = %s, type = %s (%s)", exp.loc.linnum, exp.loc.charnum, target, exp.type, exp.op);
+
+            auto source = exp.type;
+
+            // DotVarExp resolve conversions, e.g from an enum to it's base type
+            if (auto dve = exp.isDotVarExp())
+                source = dve.var.type;
+
+            if (!source)
+                // Defensively assume that the cast is reqired
```
Comment (Geod24) [inline src/dmd/dtoh.d:1888]: required

File: src/dmd/dtoh.d [hunk 4]
```diff
@@ -1855,7 +1856,107 @@ public:
...
+                    visit(ta);
+                    buf.writestring("()");
+                }
+                return;
+            }
+
+            if (auto se = exp.isStringExp())
+            {
+                // Rewrite as <length> + <literal> pair optionally
+                // wapped in a initializer list/ctor call
```
Comment (Geod24) [inline src/dmd/dtoh.d:1926]: wrapped

File: src/dmd/dtoh.d [hunk 5]
```diff
@@ -652,7 +652,8 @@ public:
 
                 case EnumKind.String, EnumKind.Enum:
                     buf.writestring("static ");
-                    writeEnumTypeName(type);
+                    auto target = determineEnumType(type);
+                    target.accept(this);
```
Comment (Geod24) [inline src/dmd/dtoh.d:656]: Why is it not a single statement like the other calls ?

File: src/dmd/dtoh.d [hunk 6]
```diff
@@ -1304,7 +1305,8 @@ public:
             else
             {
                 buf.writestring("static ");
-                writeEnumTypeName(memberType);
+                auto target = determineEnumType(memberType);
+                target.accept(this);
```
Comment (Geod24) [inline src/dmd/dtoh.d:1310]: Same question about single statement.

**Review by Geod24 — APPROVED** [review]
Just a few nits

File: src/dmd/dtoh.d (hunk 5 above)
Comment (MoonlightSentinel) [reply src/dmd/dtoh.d:656]: Because `target` is used a few lines below.

File: src/dmd/dtoh.d (hunk 6 above)
Comment (MoonlightSentinel) [reply src/dmd/dtoh.d:1310]: Used below

**Review by Geod24 — APPROVED** [review]

**Pull request merged**